[jit] Enable IValue to hold a PyObject

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32491 [jit] Enable IValue to hold a PyObject**

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Differential Revision: [D19502234](https://our.internmc.facebook.com/intern/diff/D19502234/)

zhaojuanmao committed:
[jit] Enable IValue to hold a PyObject

This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Differential Revision: [D19502234](https://our.internmc.facebook.com/intern/diff/D19502234/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ad71c57c:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ad71c57c8519c719432ab28d200aebd5686883d0).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4315885) pytorch_linux_xenial_cuda10_1_cudnn7_py3_slow_test (1/3)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89446845) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89446845))
<details>
<summary>
<code>TypeError: unsupported operand type(s) for -=: 'Retry' and 'int'</code>
</summary>

```
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 576, in send 
    r = adapter.send(request, **kwargs) 
  File "/usr/share/python-wheels/CacheControl-0.11.5-py2.py3-none-any.whl/cachecontrol/adapter.py", line 46, in send 
    resp = super(CacheControlAdapter, self).send(request, **kw) 
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/adapters.py", line 376, in send 
    timeout=timeout 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/connectionpool.py", line 610, in urlopen 
    _stacktrace=sys.exc_info()[2]) 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/util/retry.py", line 228, in increment 
    total -= 1 
TypeError: unsupported operand type(s) for -=: 'Retry' and 'int' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4315883) pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_test (2/3)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89446891) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89446891))
<details>
<summary>
<code>TypeError: unsupported operand type(s) for -=: 'Retry' and 'int'</code>
</summary>

```
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 576, in send 
    r = adapter.send(request, **kwargs) 
  File "/usr/share/python-wheels/CacheControl-0.11.5-py2.py3-none-any.whl/cachecontrol/adapter.py", line 46, in send 
    resp = super(CacheControlAdapter, self).send(request, **kw) 
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/adapters.py", line 376, in send 
    timeout=timeout 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/connectionpool.py", line 610, in urlopen 
    _stacktrace=sys.exc_info()[2]) 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/util/retry.py", line 228, in increment 
    total -= 1 
TypeError: unsupported operand type(s) for -=: 'Retry' and 'int' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4315884) pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX2_test (3/3)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89446938) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89446938))
<details>
<summary>
<code>TypeError: unsupported operand type(s) for -=: 'Retry' and 'int'</code>
</summary>

```
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 576, in send 
    r = adapter.send(request, **kwargs) 
  File "/usr/share/python-wheels/CacheControl-0.11.5-py2.py3-none-any.whl/cachecontrol/adapter.py", line 46, in send 
    resp = super(CacheControlAdapter, self).send(request, **kw) 
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/adapters.py", line 376, in send 
    timeout=timeout 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/connectionpool.py", line 610, in urlopen 
    _stacktrace=sys.exc_info()[2]) 
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/util/retry.py", line 228, in increment 
    total -= 1 
TypeError: unsupported operand type(s) for -=: 'Retry' and 'int' 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zhaojuanmao committed:
Update on "[jit] Enable IValue to hold a PyObject"


This PR enables IValue to be able to hold a pure PyObject by adding a
new enum tag, a new jit_type to denote PyObject existance in IValue and
the JIT type system. We don't and not plan to expose this to user.

This is the basic piece that enable ivalue to be adopted broader like
making RRef always hold IValue, it might also simplify some compiler
logic

Differential Revision: [D19502234](https://our.internmc.facebook.com/intern/diff/D19502234/)

[ghstack-poisoned]

zhaojuanmao commented:
test failures are not relevant
